repo/fsrepo/migrations/ipfsfetcher.go

```diff
@@ -0,0 +1,297 @@
+	iface "github.com/ipfs/interface-go-ipfs-core"
+	"github.com/ipfs/interface-go-ipfs-core/options"
+	ipath "github.com/ipfs/interface-go-ipfs-core/path"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+type IpfsFetcher struct {
+	distPath string
+	limit    int64
+	peers    []string
+
+	openOnce  sync.Once
+	closeOnce sync.Once
+	err       error
+
+	ipfs       iface.CoreAPI
+	ipfsCancel context.CancelFunc
+	ipfsCtx    context.Context
+	ipfsTmpDir string
+}
+
+// NewIpfsFetcher creates a new IpfsFetcher
+//
+// Specifying "" for distPath sets the default IPNS path.
+// Specifying 0 for fetchLimit sets the default, -1 means no limit.
+func NewIpfsFetcher(distPath string, fetchLimit int64, peers []string) *IpfsFetcher {
+	f := &IpfsFetcher{
+		limit:    defaultFetchLimit,
+		distPath: LatestIpfsDist,
+		peers:    peers,
+	}
+
+	if distPath != "" {
+		if !strings.HasPrefix(distPath, "/") {
+			distPath = "/" + distPath
+		}
+		f.distPath = distPath
+	}
+
+	if fetchLimit != 0 {
+		if fetchLimit == -1 {
+			fetchLimit = 0
+		}
+		f.limit = fetchLimit
+	}
+
+	return f
+}
+
+// Fetch attempts to fetch the file at the given path, from the distribution
+// site configured for this HttpFetcher.  Returns io.ReadCloser on success,
+// which caller must close.
+func (f *IpfsFetcher) Fetch(ctx context.Context, filePath string) (io.ReadCloser, error) {
+	// Initialize and start IPFS node on first call to Fetch, since the fetcher
+	// may be created by not used.
+	f.openOnce.Do(func() {
+		f.ipfsTmpDir, f.err = initTempNode(ctx)
+		if f.err != nil {
+			return
+		}
+
+		f.err = f.startTempNode()
+	})
+
+	fmt.Printf("Fetching with IPFS: %q\n", filePath)
+
+	if f.err != nil {
+		return nil, f.err
+	}
+
+	iPath, err := parsePath(filepath.Join(f.distPath, filePath))
+	if err != nil {
+		return nil, err
+	}
+
+	nd, err := f.ipfs.Unixfs().Get(ctx, iPath)
+	if err != nil {
+		return nil, err
+	}
+
+	fileNode, ok := nd.(files.File)
+	if !ok {
+		return nil, fmt.Errorf("%q is not a file", filePath)
+	}
+
+	if f.limit != 0 {
+		return NewLimitReadCloser(fileNode, f.limit), nil
+	}
+	return fileNode, nil
+}
+
+func (f *IpfsFetcher) Close() error {
+	f.closeOnce.Do(func() {
+		if f.ipfsCancel != nil {
+			// Tell ipfs to stop
+			f.ipfsCancel()
+
+			// Wait until ipfs is stopped
+			<-f.ipfsCtx.Done()
+		}
+
+		if f.ipfsTmpDir != "" {
+			// Remove the temp ipfs dir
+			if err := os.RemoveAll(f.ipfsTmpDir); err != nil {
+				fmt.Fprintln(os.Stderr, err)
+			}
+		}
+	})
+	return nil
+}
+
+func initTempNode(ctx context.Context) (string, error) {
+	defaultPath, err := ipfsdir.IpfsDir("")
+	if err != nil {
+		return "", err
+	}
+
+	// TODO: Is there a better way to check it plugins are loaded first?
+	err = setupPlugins(defaultPath)
+	// Need to ignore errors here because plugins may already be loaded when
+	// run from ipfs daemon.
+	if err != nil {
+		fmt.Println("Ignored plugin error:", err)
```

Comment:
Will print to stderr for now, and make it look less like an error.  If ipfsfetcher is only going to be used from within go-ipfs, then it may be better to not try to load the pluging at all.